File: examples/go/statistics.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"regexp"
	"strings"
)

var campaignRe = regexp.MustCompile(`^[a-z0-9-]{1,64}$`)

// ─────────────────────────────────────────────────────────────────────────────
//  MARKET STATISTICS — MODELS
// ─────────────────────────────────────────────────────────────────────────────

type MarketStatisticsResponse struct {
	ScopeKind        string       `json:"scopeKind"`
	ScopeKey         string       `json:"scopeKey"`
	GeneratedAt      string       `json:"generatedAt"`
	IncludedSections []string     `json:"includedSections"`
	Demand           *DemandStats `json:"demand"`
	Salary           *SalaryStats `json:"salary"`
	Experience       []Bucket     `json:"experience"`
	TopLocations     []Bucket     `json:"topLocations"`
	TopSkills        []Bucket     `json:"topSkills"`
}

type DemandStats struct {
	ActiveOffers      int          `json:"activeOffers"`
	DistinctEmployers int          `json:"distinctEmployers"`
	RemoteOffers      int          `json:"remoteOffers"`
	RemotePercentage  int          `json:"remotePercentage"`
	OfferTrend        []TrendPoint `json:"offerTrend"`
}

type TrendPoint struct {
	Period     string `json:"period"`
	OfferCount int    `json:"offerCount"`
}

type SalaryStats struct {
	Currency  string      `json:"currency"`
	Overall   *SalaryBand `json:"overall"`
	B2B       *SalaryStat `json:"b2b"`
	Permanent *SalaryStat `json:"permanent"`
}

type SalaryBand struct {
	Min    float64 `json:"min"`
	P25    float64 `json:"p25"`
	Median float64 `json:"median"`
	P75    float64 `json:"p75"`
	Max    float64 `json:"max"`
}

type SalaryStat struct {
	Median     float64 `json:"median"`
	Average    float64 `json:"average"`
	OfferCount int     `json:"offerCount"`
}

type Bucket struct {
	Label      string `json:"label"`
	OfferCount int    `json:"offerCount"`
	Percentage int    `json:"percentage"`
}

// ─────────────────────────────────────────────────────────────────────────────
//  CLIENT METHOD
// ─────────────────────────────────────────────────────────────────────────────

// GetMarketStatistics fetches labour-market statistics for the given scope.
// fields is an optional subset of sections; nil/empty returns all sections
// available for the scope.
func (c *SolidJobsClient) GetMarketStatistics(scopeKind, scopeKey, campaign string, fields []string) (*MarketStatisticsResponse, error) {
	if !campaignRe.MatchString(campaign) {
		return nil, fmt.Errorf("campaign must contain only lowercase letters, digits and hyphens (max 64 chars)")
	}

	params := url.Values{}
	params.Set("campaign", campaign)
	if len(fields) > 0 {
		params.Set("fields", strings.Join(fields, ","))
	}

	reqURL := fmt.Sprintf("%s/public-api/market-statistics/%s/%s?%s", c.BaseURL, scopeKind, scopeKey, params.Encode())

	req, err := http.NewRequest(http.MethodGet, reqURL, nil)
	if err != nil {
		return nil, fmt.Errorf("creating request: %w", err)
	}
	req.Header.Set("X-Api-Version", c.APIVersion)

	resp, err := c.doWithRetry(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP %d", resp.StatusCode)
	}

	var result MarketStatisticsResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("decoding JSON: %w", err)
	}
	return &result, nil
}

// ─────────────────────────────────────────────────────────────────────────────
//  DEMO
// ─────────────────────────────────────────────────────────────────────────────

func printBand(band *SalaryBand) {
	if band == nil {
		fmt.Println("    (no offers with a declared salary)")
		return
	}
	fmt.Printf("    min=%.0f  p25=%.0f  median=%.0f  p75=%.0f  max=%.0f\n",
		band.Min, band.P25, band.Median, band.P75, band.Max)
}

func printContractSalary(label string, stat *SalaryStat) {
	if stat == nil {
		fmt.Printf("    %s: (no precomputed data)\n", label)
		return
	}
	fmt.Printf("    %s: median=%.0f  average=%.0f  offers=%d\n", label, stat.Median, stat.Average, stat.OfferCount)
}

func printBuckets(buckets []Bucket) {
	for _, b := range buckets {
		fmt.Printf("    %-24s %5d offers  (%d%%)\n", b.Label, b.OfferCount, b.Percentage)
	}
}

func printStatistics(stats *MarketStatisticsResponse) {
	fmt.Printf("Scope:            %s / %s\n", stats.ScopeKind, stats.ScopeKey)
	fmt.Printf("Generated at:     %s\n", stats.GeneratedAt)
	fmt.Printf("Included sections: %s\n", strings.Join(stats.IncludedSections, ", "))

	if d := stats.Demand; d != nil {
		fmt.Println("\n[demand]")
		fmt.Printf("  activeOffers=%d  distinctEmployers=%d  remoteOffers=%d  remotePercentage=%d%%\n",
			d.ActiveOffers, d.DistinctEmployers, d.RemoteOffers, d.RemotePercentage)
		fmt.Println("  offerTrend (quarterly):")
		for _, point := range d.OfferTrend {
			fmt.Printf("    %s: %d offers\n", point.Period, point.OfferCount)
		}
	}

	if s := stats.Salary; s != nil {
		fmt.Printf("\n[salary] currency=%s\n", s.Currency)
		fmt.Println("  overall (live percentiles):")
		printBand(s.Overall)
		printContractSalary("b2b       ", s.B2B)
		printContractSalary("permanent ", s.Permanent)
	}

	if stats.Experience != nil {
		fmt.Println("\n[experience]")
		printBuckets(stats.Experience)
	}

	if stats.TopLocations != nil {
		fmt.Println("\n[topLocations]")
		printBuckets(stats.TopLocations)
	}

	if stats.TopSkills != nil {
		fmt.Println("\n[topSkills]")
		printBuckets(stats.TopSkills)
	}
}

func runStatisticsDemo() {
	client := NewClient()
	campaign := "go-stats"

	// 1) Full snapshot — every section available for the React specialization
	full, err := client.GetMarketStatistics("subcategory", "React", campaign, nil)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
	printStatistics(full)

	// 2) Partial fetch — ask only for `demand` and `salary` using the `fields` filter.
	//    IncludedSections in the response confirms exactly what came back.
	fmt.Println("\n───────────────────────────────────────────────")
	fmt.Println("Partial fetch with fields=demand,salary:")
	partial, err := client.GetMarketStatistics("subcategory", "React", campaign, []string{"demand", "salary"})
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Included sections: %s\n", strings.Join(partial.IncludedSections, ", "))
}

File: examples/rust/src/statistics.rs
use crate::client::SolidJobsClient;
use crate::models::{Bucket, MarketStatisticsResponse, SalaryBand, SalaryStat};

fn print_band(band: &Option<SalaryBand>) {
    match band {
        None => println!("    (no offers with a declared salary)"),
        Some(b) => println!(
            "    min={:.0}  p25={:.0}  median={:.0}  p75={:.0}  max={:.0}",
            b.min, b.p25, b.median, b.p75, b.max
        ),
    }
}

fn print_contract_salary(label: &str, stat: &Option<SalaryStat>) {
    match stat {
        None => println!("    {}: (no precomputed data)", label),
        Some(s) => println!(
            "    {}: median={:.0}  average={:.0}  offers={}",
            label, s.median, s.average, s.offer_count
        ),
    }
}

fn print_buckets(buckets: &[Bucket]) {
    for b in buckets {
        println!("    {:<24} {:>5} offers  ({}%)", b.label, b.offer_count, b.percentage);
    }
}

fn print_statistics(stats: &MarketStatisticsResponse) {
    println!("Scope:            {} / {}", stats.scope_kind, stats.scope_key);
    println!("Generated at:     {}", stats.generated_at);
    println!("Included sections: {}", stats.included_sections.join(", "));

    if let Some(d) = &stats.demand {
        println!("\n[demand]");
        println!(
            "  activeOffers={}  distinctEmployers={}  remoteOffers={}  remotePercentage={}%",
            d.active_offers, d.distinct_employers, d.remote_offers, d.remote_percentage
        );
        println!("  offerTrend (quarterly):");
        for point in &d.offer_trend {
            println!("    {}: {} offers", point.period, point.offer_count);
        }
    }

    if let Some(s) = &stats.salary {
        println!("\n[salary] currency={}", s.currency);
        println!("  overall (live percentiles):");
        print_band(&s.overall);
        print_contract_salary("b2b       ", &s.b2b);
        print_contract_salary("permanent ", &s.permanent);
    }

    if let Some(experience) = &stats.experience {
        println!("\n[experience]");
        print_buckets(experience);
    }

    if let Some(top_locations) = &stats.top_locations {
        println!("\n[topLocations]");
        print_buckets(top_locations);
    }

    if let Some(top_skills) = &stats.top_skills {
        println!("\n[topSkills]");
        print_buckets(top_skills);
    }
}

pub async fn run() -> Result<(), Box<dyn std::error::Error>> {
    let client = SolidJobsClient::new();
    let campaign = "rust-stats";

    // 1) Full snapshot — every section available for the React specialization
    let full = client
        .get_market_statistics("subcategory", "React", campaign, &[])
        .await?;
    print_statistics(&full);

    // 2) Partial fetch — ask only for `demand` and `salary` using the `fields` filter.
    //    included_sections in the response confirms exactly what came back.
    println!("\n───────────────────────────────────────────────");
    println!("Partial fetch with fields=demand,salary:");
    let partial = client
        .get_market_statistics("subcategory", "React", campaign, &["demand", "salary"])
        .await?;
    println!("Included sections: {}", partial.included_sections.join(", "));

    Ok(())
}
